### ISSUE TITLE
Feature request: gfor

### ISSUE THREAD

**Issue by ngphuoc** [JuliaComputing/ArrayFire.jl#184]
It would be nice to have gfor in julia.
References:
http://arrayfire.org/docs/page_gfor.htm
https://groups.google.com/forum/#!topic/arrayfire-users/t-vmSq-wtm8

**Comment by gaika** [comment]
`gfor` is implemented via broadcast, should work out of the box. Just use `.= `, `.+`, `.()`, etc. and it should work, if it doesn't, file a bug.

**Comment by ngphuoc** [comment]
I am just curious. Could we have explicit gfor in julia? I see arrayfire allows user function but it's in C++. Is it possible to compile user functions in julia to C++ and feed to gfor?

**Comment by gaika** [comment]
gfor is not a function:
```
#define gfor(var, ...) for (var = af::seq(af::seq(__VA_ARGS__), true); af::gforToggle(); )
```

**Comment by constantza** [comment]
Hello,
I'm doing GPU Computing. I have seen in the past similar attitude against the GFOR.
Just for answering to ngphuoc. I'm not Julia developer, nor ArrayFire developer.
I use GFOR constructor in another environment similar to ArrayFire, but in Matlab.
I'm thinking to migrate in JULIA, but unfortunately, no GFOR here.

This question asked by ngphuoc was reiterated in the past. And systematically neglected following simplistic considerations...

As an example, (https://discourse.julialang.org/t/arrayfire-julia-implementation-of-gfor/633)
the previous developer of the ArrayFire (aka Ranjan) replied to someone asking for GFOR implementation under Julia:
"So here’s the thing: the gfor in the arrayfire library is simply syntactic sugar for the vector operations.
For example, something like
@gfor for i = 1:10
a[i] = a[i] + 1
end
is essentially the same as a = a + 1."

What kind of strange ignorance about parallel computing, sorry to precise it.... surprising from people which proved to be really smart in computer science.

May you imagine a lot of jobs (thousands or millions obtained after a granular work on some very big object? The interest for this is galactic more important than simulating the addition of two vectors or any other broadcasting operators (by default supported natively in Julia, so why citing them?) ? When the jobs are highly nonlinear, is impossible to reduce all to simple combination of standard operators. Obvious.

For example any very heavy nonparametric or parametric Monte Carlo or statistical bootstrap necessary for the inference. People use Python as glue language because is able to do this under Theano. 

ArrayFire is an opportunity to implement in Julia parallel Jobs directly on the GPU, that it. 
Is not for people which want to add vectors, or reduce the parallel jobs to standard simple operations between vectors.

So in conclusion, YES. Julia needs the implementation of GFOR from ArrayFire. Even Nim (less developed than JULIA, and more archaic) has its GFOR implementation. Why not JULIA?  

And of course, GFOR is not a function :-). But some constructor able to parallelize nonlinear or linear jobs working on granular data.

Best regards,
constantza

**Comment by SimonDanisch** [comment]
You might want to have a look at [GPUArrays](https://github.com/JuliaGPU/GPUArrays.jl) which actually allows you to compile quite arbitrary Julia Code for the GPU.

I wrote up a relatively simple [introduction](https://medium.com/@sdanisch/writing-extendable-and-hardware-agnostic-gpu-libraries-b21c145a8dad).
Concrete array packages are [CLArrays](https://github.com/JuliaGPU/CLArrays.jl) and [CuArrays](https://github.com/JuliaGPU/CuArrays.jl) (which both inherit from GPUArrays, so support the same operations).

You can implement pretty complicated, platform independent GPU kernels with those in pure Julia, but also have the easy ArrayFire like vector operations.

**Comment by pavanky** [comment]
As someone who had a significant hand in developing the current version of arrayfire (and a developer of Jacket for MATLAB before this) for many years, I have to say adding GFOR is never needed. 

It was a good hand holding tool for *beginners* who were using for loops in maltab when they really shouldn't have (i.e. using loops instead of broadcasting or vectorizing). GFOR took cases like this and autovectorized them in the backend. Unfortunately this created the illusion that GFOR supported all kinds of for loops (for example accumulation wasn't / isn't supported). Any operations done inside GFOR were *required* to be independent of each other. 

When arrayfire 3.0 rolled around, the C API for arrayfire was modified to perform vectorizations and broadcasting without having to use GFOR, and using a much smaller memory footprints. 

The sad reality of GFOR is that people expect it to this magical thing when it is not. I have seen many bug reports of when people were trying to use it for something it was never capable of doing. I am not sure if we did a bad job explaining the limitations of GFOR or did a really good job marketing its usefulness.
